tests/unit/Db/FailedLinkAccessMapperTest.php

```diff
@@ -0,0 +1,170 @@
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\IDBConnection;
+use Test\TestCase;
+
+/**
+ * @group DB
+ */
+class FailedLinkAccessMapperTest extends TestCase {
+
+	/** @var  FailedLinkAccess $mapper*/
+	private $mapper;
+
+	/** @var  IDBConnection $connection*/
+	private $connection;
+
+	/** @var BruteForceProtectionConfig | \PHPUnit\Framework\MockObject\MockObject $configMock */
+	private $configMock;
+
+	/** @var ITimeFactory | \PHPUnit\Framework\MockObject\MockObject $timeFactory */
+	private $timeFactoryMock;
+
+	/** @var int $baseTime */
+	private $baseTime;
+
+	/** @var string  */
+	private $dbTable = 'bfp_link_accesses';
+
+	/** @var int $thresholdConfigVal */
+	private $thresholdConfigVal = 60;
+
+	public function setUp(): void {
+		parent::setUp();
+		$this->baseTime = \time();
+		$this->timeFactoryMock = $this->createMock(ITimeFactory::class);
+		$this->configMock = $this->createMock(BruteForceProtectionConfig::class);
+		$this->connection = \OC::$server->getDatabaseConnection();
+		$this->mapper = new FailedLinkAccessMapper($this->connection, $this->configMock, $this->timeFactoryMock);
+
+		$query = $this->connection->getQueryBuilder()->select('*')->from($this->dbTable);
+		$result = $query->execute()->fetchAll();
+		$this->assertEmpty($result, 'we need to start with a empty bfp_link_accesses table');
+
+		$this->addInitialTestEntries();
+	}
+
+	public function addInitialTestEntries() {
+		$linkAccess = new FailedLinkAccess();
+		$linkAccess->setLinkToken('token1');
+		$linkAccess->setIp("192.168.1.1");
+		$linkAccess->setAttemptedAt($this->baseTime+20);
+		$this->mapper->insert($linkAccess);
+
+		$linkAccess = new FailedLinkAccess();
+		$linkAccess->setLinkToken('token1');
+		$linkAccess->setIp("192.168.1.1");
+		$linkAccess->setAttemptedAt($this->baseTime+60);
+		$this->mapper->insert($linkAccess);
+
+		$linkAccess = new FailedLinkAccess();
+		$linkAccess->setLinkToken('token1');
+		$linkAccess->setIp("192.168.1.2");
+		$linkAccess->setAttemptedAt($this->baseTime+60);
+		$this->mapper->insert($linkAccess);
+
+		$linkAccess = new FailedLinkAccess();
+		$linkAccess->setLinkToken('token2');
+		$linkAccess->setIp("192.168.1.1");
+		$linkAccess->setAttemptedAt($this->baseTime+60);
+		$this->mapper->insert($linkAccess);
+
+		$linkAccess = new FailedLinkAccess();
+		$linkAccess->setLinkToken('token1');
+		$linkAccess->setIp("192.168.1.1");
+		$linkAccess->setAttemptedAt($this->baseTime+100);
+		$this->mapper->insert($linkAccess);
+	}
+
+	public function tearDown(): void {
+		parent::tearDown();
+		$query = $this->connection->getQueryBuilder()->delete($this->dbTable);
+		$query->execute();
+	}
+
+	public function testGetFailedAccessCountForTokenIpCombination() {
+		$functionCallTime = $this->baseTime+110;
+		$this->configMock->expects($this->exactly(3))
+			->method('getBruteForceProtectionTimeThreshold')
+			->willReturn($this->thresholdConfigVal);
+		$this->timeFactoryMock->expects($this->exactly(3))
```

Comment:
If we need to verify the number of calls to a specific method, I'd rather check it per method run, not after having run the method several times.
I'd recommend to either remove the check or use a data provider

Reply:
Verifying the number of calls is not the aim of this test. The check removed.